### ISSUE TITLE
[MPMD-290] Add CPD example for C#

### ISSUE THREAD

**Issue by jira-importer** [apache/maven-pmd-plugin#623]
**[QualiteSys QualiteSys](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=qualitesys)** opened **[MPMD-290](https://issues.apache.org/jira/browse/MPMD-290?redirect=false)** and commented

I wish to analyze C# source code with CPD, with no success

Looking to the CpdReport.java MOJO, it seams that the language is not extracted from either the pom.xml file or the CLI options.

Line 76 :

    `@Parameter`( defaultValue = "java" )
    private String language;

property = "language" is missing, so the language is always forced to java

Is it possible to fix this bug, and add the missing languages CPD is supporting ?

Thanks

 

---

**Affects:** 3.12.0

**Issue Links:**
- [MPMD-221](https://issues.apache.org/jira/browse/MPMD-221) Support all languages that PMD supports

**Remote Links:**
- [GitHub Pull Request #32
](https://github.com/apache/maven-pmd-plugin/pull/32)

**Comment by jira-importer** [comment]
**[QualiteSys QualiteSys](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=qualitesys)** commented

Any news on this issue ?

**Comment by jira-importer** [comment]
**[Andreas Dangel](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=adangel)** commented

As documented here [https://maven.apache.org/plugins/maven-pmd-plugin/cpd-mojo.html#language](https://maven.apache.org/plugins/maven-pmd-plugin/cpd-mojo.html#language):

> Valid values are currently java, javascript or jsp.

Looking at the code, I don't see a reason, why it shouldn't work with other languages. The language is used here: https://github.com/apache/maven-pmd-plugin/blob/829ea8e4292d50153158df8a93794025b62edfca/src/main/java/org/apache/maven/plugins/pmd/CpdReport.java#L284

Have you tried to set the language to "cs" (that's the language code for c# for CPD)?
How does your pom.xml look like? Did you also configure the [includes](https://maven.apache.org/plugins/maven-pmd-plugin/cpd-mojo.html#includes) property? It defaults do "**/*.java", so by default doesn't consider any "**/*.cs" files...

I'll try to add a integration test, to verify whether this works or doesn't work....

**Comment by jira-importer** [comment]
**[Andreas Dangel](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=adangel)** commented

So, it works with the following snippet:

```java
      <plugin>
        <groupId>org.apache.maven.plugins</groupId>
        <artifactId>maven-pmd-plugin</artifactId>
        <version>3.13.0</version>
        <configuration>
          <language>cs</language>
          <minimumTokens>10</minimumTokens>
          <includes>
            <include>**/*.cs</include>
          </includes>
          <compileSourceRoots>
            <compileSourceRoot>${basedir}/src/main/cs</compileSourceRoot>
          </compileSourceRoots>
          <printFailingErrors>true</printFailingErrors>
        </configuration>
        <executions>
          <execution>
            <goals>
              <goal>cpd-check</goal>
            </goals>
          </execution>
        </executions>
        <dependencies>
            <dependency>
                <groupId>net.sourceforge.pmd</groupId>
                <artifactId>pmd-cs</artifactId>
                <version>6.21.0</version>
            </dependency>
        </dependencies>
      </plugin>
 
```

 

The additional dependency is necessary, so that the correct tokenizer is used. Otherwise comments are not ignored and might be reported as duplicates.

I don't know, how this configuration interacts with other plugins, e.g. dotnet-maven-plugin.

So maybe, it's just a documentation issue in the end?

**Comment by jira-importer** [comment]
**[Andreas Dangel](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=adangel)** commented

The PR https://github.com/apache/maven-pmd-plugin/pull/32 adds an integration test and also a example configuration in the documentation.

The original problem is not reproducible.

**Comment by jira-importer** [comment]
**[Andreas Dangel](https://issues.apache.org/jira/secure/ViewProfile.jspa?name=adangel)** commented

Fixed with [597d4cb712](https://gitbox.apache.org/repos/asf?p=maven-pmd-plugin.git;a=commit;h=597d4cb712cf63a17daf67e6f0c0516fe1643e69)
